sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroSerializer.ts

```diff
@@ -0,0 +1,226 @@
+//       the purpose of this serialzer to move the schema out of the message
+//       payload and into the schema registry.
+//
+const FORMAT_INDICATOR = 0;
+const SCHEMA_ID_OFFSET = 4;
+const SCHEMA_ID_LENGTH = 32;
+const PAYLOAD_OFFSET = 36;
+
+interface CacheEntry {
+  /** Schema ID */
+  id: string;
+
+  /** avsc-specific representation for schema */
+  type: avro.Type;
+}
+
+/**
+ * Options for Schema
+ */
+export interface SchemaRegistryAvroSerializerOptions {
+  /**
+   * When true, register new schemas passed to serialize. Otherwise, and by
+   * default, fail if schema has not already been registered.
+   *
+   * Automatic schema registration is NOT recommended for production scenarios.
+   */
+  autoRegisterSchemas?: boolean;
+}
+
+/**
+ * Avro serializer that obtains schemas from a schema registry and does not
+ * pack schemas into its payloads.
+ */
+export class SchemaRegistryAvroSerializer {
+  /**
+   * Creates a new serializer.
+   *
+   * @param registry Schema Registry where schemas are registered and obtained.
+   *                 Usually this is a SchemaRegistryClient instance.
+   *
+   * @param schemaGroup The schema group to use when making requests to the
+   *                    registry.
+   */
+  constructor(
+    registry: SchemaRegistry,
+    schemaGroup: string,
+    options?: SchemaRegistryAvroSerializerOptions
+  ) {
+    this.registry = registry;
+    this.schemaGroup = schemaGroup;
+    this.autoRegisterSchemas = options?.autoRegisterSchemas ?? false;
+  }
+
+  /** The schema group to use when making requests to the registry*/
+  public readonly schemaGroup: string;
+
+  /**
+   * Schema Registry where schemas are registered and obtained.
+   * Usually this is a SchemaRegistryClient instance.
+   */
+  public readonly registry: SchemaRegistry;
+
+  /**
+   * When true, register new schemas passed to serialize. Otherwise, and by
+   * default, fail if schema has not already been registered.
+   *
+   * Automatic schema registration is NOT recommended for production scenarios.
+   */
+  public autoRegisterSchemas: boolean;
+
+  // REVIEW: signature.
+  //
+  // - Better to serialize into a stream? I aborted that for now as I wanted to
+  //   do the simplest thing that could possibly work first to make sure there
+  //   were no blockers in our dependencies. I also wanted to get feedback on
+  //   what the API shape should be before diving into that.
+  //
+  // - This type should ultimately be able to implement a core ObjectSerializer
+  //   interface. Do we know what that would look like? Maybe it takes `any` as
+  //   the format-specific schema/type info? Or does it always take a
+  //   format-specific schema string?
+  //
+  //   The C#/Java approach of passing Type and assuming every serializer can
+  //   get its schema by reflecting on the type does not work for JavaScript. We
+  //   need to support arbitrary objects that match a schema.
```

Comment:
this is something I actually did talk to @bterlson about before. I think our contract has to be a little different since we can't do the reflection trick. 

Could we make `ObjectSerializer` generic over the Schema? e.g.

```ts

interface ObjectSerializer<SchemaType> {
   serialize(value: any, schema: SchemaType): Promise<Buffer>;
}
```